File: telegram-bot-core/src/main/java/ru/tggc/telegrambotcore/util/ParamConverter.java
package ru.tggc.telegrambotcore.util;

import lombok.experimental.UtilityClass;

@UtilityClass
public class ParamConverter {

    @SuppressWarnings("unchecked")
    public Object convert(String value, Class<?> targetType) {
        if (value == null) {
            return null;
        }
        if (targetType.equals(String.class)) {
            return value;
        } else if (targetType.equals(int.class) || targetType.equals(Integer.class)) {
            return Integer.parseInt(value);
        } else if (targetType.equals(long.class) || targetType.equals(Long.class)) {
            return Long.parseLong(value);
        } else if (targetType.equals(boolean.class) || targetType.equals(Boolean.class)) {
            return Boolean.parseBoolean(value);
        } else if (Enum.class.isAssignableFrom(targetType)) {
            return Enum.valueOf((Class<Enum>) targetType, value.toUpperCase());
        }

        throw new IllegalArgumentException("Unsupported parameter type: " + targetType.getName());
    }
}

File: telegram-bot-core/src/main/kotlin/ru/tggc/telegrambotcore/util/ParamConverter.kt
package ru.tggc.telegrambotcore.util

class ParamConverter {
    companion object {

        fun convert(value: String?, targetType: Class<*>): Any? {
            if (value == null) {
                return null
            }

            return when {
                targetType == String::class.java -> value

                targetType == Int::class.java || targetType == Int::class.javaObjectType -> value.toInt()

                targetType == Long::class.java || targetType == Long::class.javaObjectType -> value.toLong()

                targetType == Boolean::class.java || targetType == Boolean::class.javaObjectType -> value.toBoolean()

                targetType.isEnum -> targetType.enumConstants.first { (it as Enum<*>).name == value.uppercase() }

                else -> throw IllegalArgumentException("Unsupported parameter type: ${targetType.name}")
            }
        }
    }
}
